fix(ivy): properly parenthesize ternary expressions when emitted

alxhub opened the pull request:
Previously, ternary expressions were emitted as:

condExpr ? trueCase : falseCase

However, this causes problems when ternary operations are nested. In
particular, a template expression of the form:

a?.b ? c : d

would have compiled to:

a == null ? null : a.b ? c : d

The precedence here is incorrect:

a == null ? null : (a.b ? c : d)

when in reality the desired association is:

(a == null ? null : a.b) ? c : d

This commit adds parentheses when emitting ternary operations to preserve
the desired precedence.

AndrewKushnir reviewed (approved):
LGTM 👍 Thanks for the fix Alex!

kara reviewed (approved):
LGTM too, though looks like one more compliance test needs to be updated (CI is unhappy)

alxhub committed:
fix(ivy): properly parenthesize ternary expressions when emitted

Previously, ternary expressions were emitted as:

condExpr ? trueCase : falseCase

However, this causes problems when ternary operations are nested. In
particular, a template expression of the form:

a?.b ? c : d

would have compiled to:

a == null ? null : a.b ? c : d

The ternary operator is right-associative, so that expression is interpreted
as:

a == null ? null : (a.b ? c : d)

when in reality left-associativity is desired in this particular instance:

(a == null ? null : a.b) ? c : d

This commit adds a check in the expression translator to detect such
left-associative usages of ternaries and to enforce such associativity with
parentheses when necessary.

A test is also added for the template type-checking expression translator,
to ensure it correctly produces right-associative expressions for ternaries
in the user's template.

Fixes #34087

AndrewKushnir reviewed (approved):
👍(after the latest set of changes)

AndrewKushnir commented:
[Presubmit](http://test/OCL:284228209:BASE:284228253:1575661085709:6916b13a)
